gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/oauth2/OAuth2SsoConfig.groovy

```diff
@@ -0,0 +1,137 @@
+import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken
+
+/**
+ * Each time a class extends WebSecurityConfigurerAdapter, a new set of matchers + filters gets added
+ * to the FilterChainProxy. This is not great when we want to protect the same URLs with different kinds of
+ * security mechanisms, because only the first set of filters that hit the matchers will be executed.
+ *
+ * In order to get around this, we must only have 1 class that extends WebSecurityConfigurerAdapter.
+ * Unfortunately, the OAuth2 library (sping-security-oauth) has such a class baked in. If we have OAuth2 enabled,
+ * AND want an additional kind of security mechanism (X509, for example), we must use the
+ * {@link OAuth2SsoConfigurerAdapter} hook the library has provided. This way, only a single set of matchers
+ * and filters are added.
+ */
+@Configuration
+@SpinnakerAuthConfig
+// Use @EnableWebSecurity if/when updated to Spring Security 4.
+@EnableWebMvcSecurity
+@Import(SecurityAutoConfiguration)
+@EnableOAuth2Sso
+// Note the 4 single-quotes below - this is a raw groovy string, because SpEL and groovy
+// string syntax overlap!
+@ConditionalOnExpression(''''${spring.oauth2.client.clientId:}'!=""''')
+class OAuth2SsoConfig extends OAuth2SsoConfigurerAdapter {
+
+  @Override
+  void match(OAuth2SsoConfigurer.RequestMatchers matchers) {
+    matchers.antMatchers('/**')
+  }
+
+  @Override
+  void configure(HttpSecurity http) throws Exception {
+    AuthConfig.configure(http)
+  }
+
+  /**
+   * ResourceServerTokenServices is an interface used to manage access tokens. The UserInfoTokenService object is an
+   * implementation of that interface that uses an access token to get the logged in user's data (such as email or
+   * profile). We want to customize the Authentication object that is returned to include our custom (Kork) User.
+   */
+  @Primary
+  @Bean
+  ResourceServerTokenServices spinnakerAuthorityInjectedUserInfoTokenServices() {
+    return new ResourceServerTokenServices() {
+
+      @Autowired
+      ResourceServerProperties sso
+
+      @Autowired
+      UserInfoTokenServices userInfoTokenServices
+
+      @Autowired
+      AccountsService accountsService
+
+      @Autowired
+      UserRolesProvider userRolesProvider
+
+      @Override
+      OAuth2Authentication loadAuthentication(String accessToken) throws AuthenticationException, InvalidTokenException {
+        OAuth2Authentication oAuth2Authentication = userInfoTokenServices.loadAuthentication(accessToken)
+
+        Map details = oAuth2Authentication.userAuthentication.details as Map
+        def email = details.email ?: details.userPrincipalName
+        def roles = userRolesProvider.loadRoles(email)
+
+        // TODO(ttomsu): Pull these values into config so that every OAuth provider doesn't need to specify how to
+        // extract user details.
+        User spinnakerUser = new User(
```

Comment:
The response you've pasted in is the response from Gate, not from Github. It is the serialized version of [User](https://github.com/spinnaker/kork/blob/master/kork-security/src/main/groovy/com/netflix/spinnaker/security/User.groovy). 

The Github response looks something like [this](https://developer.github.com/v3/users/#response-1):
```
{
  "login": "octocat",
  "id": 1,
  "avatar_url": "https://github.com/images/error/octocat_happy.gif",
  "gravatar_id": "",
  "url": "https://api.github.com/users/octocat",
  "html_url": "https://github.com/octocat",
  "followers_url": "https://api.github.com/users/octocat/followers",
  "following_url": "https://api.github.com/users/octocat/following{/other_user}",
  "gists_url": "https://api.github.com/users/octocat/gists{/gist_id}",
  "starred_url": "https://api.github.com/users/octocat/starred{/owner}{/repo}",
  "subscriptions_url": "https://api.github.com/users/octocat/subscriptions",
  "organizations_url": "https://api.github.com/users/octocat/orgs",
  "repos_url": "https://api.github.com/users/octocat/repos",
  "events_url": "https://api.github.com/users/octocat/events{/privacy}",
  "received_events_url": "https://api.github.com/users/octocat/received_events",
  "type": "User",
  "site_admin": false,
  "name": "monalisa octocat",
  "company": "GitHub",
  "blog": "https://github.com/blog",
  "location": "San Francisco",
  "email": "octocat@github.com",
  "hireable": false,
  "bio": "There once was...",
  "public_repos": 2,
  "public_gists": 1,
  "followers": 20,
  "following": 0,
  "created_at": "2008-01-14T04:33:35Z",
  "updated_at": "2008-01-14T04:33:35Z",
  "total_private_repos": 100,
  "owned_private_repos": 100,
  "private_gists": 81,
  "disk_usage": 10000,
  "collaborators": 8,
  "plan": {
    "name": "Medium",
    "space": 400,
    "private_repos": 20,
    "collaborators": 0
  }
}
```

As for making this configurable, I imagine it would look something like how the [SamlSsoConfig](https://github.com/spinnaker/gate/blob/google-oauth/gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/saml/SamlSsoConfig.groovy#L79) does it.